misc/testerina/modules/testerina-core/src/main/ballerina/execute.bal

```diff
@@ -36,56 +40,107 @@ public function startSuite() returns int {
     } else {
         if testRegistry.getFunctions().length() == 0 && testRegistry.getDependentFunctions().length() == 0 {
             println("\tNo tests found");
-            return exitCode;
+            lock {
+                return exitCode;
+            }
+
         }
 
         error? err = orderTests();
         if err is error {
-            exitCode = 1;
+            enableExit();
             println(err.message());
         } else {
             executeBeforeSuiteFunctions();
-            executeTests();
+            err = executeTests();
+            if err is error {
+                enableExit();
+                println(err.message());
+            }
             executeAfterSuiteFunctions();
             reportGenerators.forEach(reportGen => reportGen(reportData));
         }
     }
-    return exitCode;
+    lock {
+        return exitCode;
+    }
 }
 
-function executeTests() {
+function executeTests() returns error? {
+    decimal startTime = currentTimeInMillis();
     foreach TestFunction testFunction in testRegistry.getFunctions() {
-        executeTest(testFunction);
+        _ = testFunction.parallelizable ? conMgr.addInitialParallelTest(testFunction) : conMgr.addInitialSerialTest(testFunction);
+    }
+
+    while !conMgr.isExecutionDone() {
+
+        if conMgr.getAvailableWorkers() != 0 {
+            conMgr.waitUntilEmptyQueueFilled();
+
+            if conMgr.getSerialQueueLength() != 0 && conMgr.getAvailableWorkers() == testWorkers {
+                TestFunction testFunction = conMgr.getSerialTest();
+                conMgr.addTestInExecution(testFunction);
+                conMgr.allocateWorker();
+                future<error?> serialWaiter = start executeTest(testFunction);
+                any _ = check wait serialWaiter;
```

Comment:
Why do you do an async call here?

Reply:
Tests which are marked by user as serialTests should be run in serial. So, here I am waiting for the results of each test before starting it.

Reply:
Why can't you directly call the function/executeTest?

Reply:
It seems I can call  it directly. I'll think a bit